lib/disk/modules/VMWareSparseDisk.rb

```diff
@@ -2,49 +2,117 @@
+
+    # 4 entries are used to detect when an extent file has been corrupted by transferring it using FTP in text mode.
+    #   The entries should be initialized with the following values:
+    'a1', 'singleEndLineChar',    # should be '\n'
+    'a1', 'nonEndLineChar',       # should be ' '
+    'a1', 'doubleEndLineChar1',   # should be '\r'
+    'a1', 'doubleEndLineChar2',   # should be '\n'
+
+    # describes the type of compression used to compress every grain in the virtual disk.
+    #   If bit 16 of the field flags is not set, compressAlgorithm is COMPRESSION_NONE.
+    #   The deflate algorithm is described in RFC 1951.
+    'S',  'compressAlgorithm',
+
+    # 'a433', 'padding'
   ])
   SIZEOF_SPARSE_EXTENT_HEADER = SPARSE_EXTENT_HEADER.size
-  GTE_SIZE = 4
-  GDE_SIZE = 4
+
+  SPARSE_EXTENT_COMPRESSED_GRAIN_HEADER = BinaryStruct.new([
+    'Q',  'lba',     # offset in the virtual disk where the block of compressed data is located
+    'L',  'size',    # size of the compressed data in bytes
+                     # the rest is data compressed with RFC 1951
+  ])
+  SIZEOF_SPARSE_EXTENT_COMPRESSED_GRAIN_HEADER = SPARSE_EXTENT_COMPRESSED_GRAIN_HEADER.size
+
+  BYTES_PER_SECTOR = 512
+  GDE_SIZE         = 4
+  GDES_PER_GD      = BYTES_PER_SECTOR / GDE_SIZE
+  GD_AT_END        = 0xFFFFFFFFFFFFFFFF
+  GTE_SIZE         = 4
+  GTES_PER_GT      = 512
+
+  COMPRESSION_NONE    = 0
+  COMPRESSION_DEFLATE = 1
+
+  FLAG_MASK_VALID_NEWLINE_DETECTION_TEST = 0x00010000
+  FLAG_MASK_USE_REDUNDANT_GRAIN_TABLE    = 0x00020000
+  FLAG_MASK_USE_ZEROED_GRAIN_GTE         = 0x00040000
+  FLAG_MASK_COMPRESSED                   = 0x00000001
+  FLAG_MASK_METADATA_MARKERS_USED        = 0x00000002
 
   def d_init
     self.diskType = "VMWare Sparse"
-    self.blockSize = 512
+    self.blockSize = BYTES_PER_SECTOR
     fileMode = MiqDiskCommon.file_mode(dInfo)
     @vmwareSparseDisk_file = MiqLargeFile.open(dInfo.fileName, fileMode)
     buf = @vmwareSparseDisk_file.read(SIZEOF_SPARSE_EXTENT_HEADER)
     @sparseHeader = OpenStruct.new(SPARSE_EXTENT_HEADER.decode(buf))
-    @grainSize = @sparseHeader.grainSize
-    @grainBytes = @grainSize * blockSize
-    @capacity = @sparseHeader.capacity * blockSize
 
     #
-    # Seek to start of the grain directory.
+    # If the grain directory sector number value is GD_AT_END and the "use redundant grain table" flag is set,
+    #  then the SPARSE_EXTENT_HEADER is located in the footer.
+    # The footer is a secondary file header stored at offset -1024 relative from the end of the file (stream)
+    #  that contains the correct grain directory sector number value.
     #
-    @vmwareSparseDisk_file.seek(@sparseHeader.gdOffset * blockSize, IO::SEEK_SET)
+    if flag_set?(@sparseHeader.flags, FLAG_MASK_USE_REDUNDANT_GRAIN_TABLE) && (GD_AT_END == @sparseHeader.gdOffset)
+      @vmwareSparseDisk_file.seek(-2 * BYTES_PER_SECTOR, IO::SEEK_END)
```

Comment:
> `.seek(-2 * BYTES_PER_SECTOR, IO::SEEK_END)`

I think this should be `.seek(-SIZEOF_SPARSE_EXTENT_HEADER)` since next you `.read(SIZEOF_SPARSE_EXTENT_HEADER)`, it just happens to be that the sparse extent header is 2 sectors

Reply:
I can add the padding if you think it will improve readability.

Reply:
Also, I thought I explained it in the comments above (derived from documentation)